api/src/main/java/org/openmrs/obs/handler/AbstractHandler.java

```diff
@@ -56,49 +59,102 @@ public AbstractHandler() {
+			// In this case, there is no extension
 			String now = longfmt.format(new Date());
-			outputfile = new File(dir, now);
+			filename += now;
 		} else {
-			title = title.replace("." + extension, "");
-			outputfile = new File(dir, title + "." + extension);
+			// If a title exists, first add the title part without its extension
+			extension = "." + getExtension(title);
+			String extensionCleanedTitle = title.replace(extension, "");
+			filename += extensionCleanedTitle;
+			
+			if (extensionCleanedTitle.equals(title)) {
+				// There was no extension if the cleaned title is the same as before
+				extension = "";
+			}
 		}
 		
+		String mimeType = obs.getComplexData().getMimeType();
+		
+		if (StringUtils.isBlank(mimeType)) {
+			// Try to get MIME type from the extension
+			MimetypesFileTypeMap fileTypeMap = new MimetypesFileTypeMap();  
+			mimeType = fileTypeMap.getContentType(title);  
+			log.debug("Probed MIME type as: " + mimeType);	
+		}
+		
+		if (StringUtils.isNotBlank(mimeType)) {
+			// Add the mimetype to the file name if it is not blank
+			mimeType = "_" + mimeType.replace("/", "-slash-");
+			filename += mimeType;
+		}
+		
+		String filenameWithExtension = filename + extension; // extension is a blank String if no extension was found
+		
+		File dir = OpenmrsUtil.getDirectoryInApplicationDataDirectory(Context.getAdministrationService().getGlobalProperty(
+		    OpenmrsConstants.GLOBAL_PROPERTY_COMPLEX_OBS_DIR));
+		
+		File outputfile = new File(dir, filenameWithExtension);
+		
 		int i = 0;
-		String tmp = null;
 		
 		// If the Obs does not exist, but the File does, append a two-digit
 		// count number to the filename and save it.
 		while (obs.getObsId() == null && outputfile.exists() && i < 100) {
-			tmp = null;
-			// Remove the extension from the filename.
-			tmp = String.valueOf(outputfile.getAbsolutePath().replace("." + extension, ""));
 			outputfile = null;
 			// Append two-digit count number to the filename.
-			String filename = (i < 1) ? tmp + "_" + nf.format(Integer.valueOf(++i)) : tmp.replace(nf.format(Integer
-			        .valueOf(i)), nf.format(Integer.valueOf(++i)));
+			String filenameCount = "_" + nf.format(Integer.valueOf(++i));
 			// Append the extension to the filename.
-			outputfile = new File(filename + "." + extension);
+			outputfile = new File(dir, filename + filenameCount + extension);
 		}
 		
 		return outputfile;
 		
 	}
 	
+	/**
+	 * Reads the mime type from the file name or probes it if that is not possible.
+	 * 
+	 * @param file the file to get the mimetype from
+	 * @return the file's mimetype
+	 */
+	public String getMimeTypeFromFile(File file) {
+		if (file == null) {
+			return null;
+		}
+		
+		String filenameWithoutExtension = FilenameUtils.removeExtension(file.getName());
+		
+		String mimeType = null;
+		// Try to parse MIME type from filename
+		if (filenameWithoutExtension.contains("-slash-")) {
+			String[] fileParams = filenameWithoutExtension.split("_");
```

Comment:
Do you have an example where MimetypesFileTypeMap in the else clause will not be enough to get the mime type? In other words, if we dropped the adding of "-slash-mimetype" to the file name, shouldn't MimetypesFileTypeMap.getContentType(file.getName()) be enough for do what we need?

Reply:
This is pretty much what I've meant also when we talked on IRC, when I said we can simply replace some lines instead of what the ticket suggested.

It's a bit of a complicated story - At first, tmarzeion suggested we probe the mimetype from the file content even if no file extension exists. The problem with this is, as you have seen, that it is inconsistent across different systems. That is why I've replaced it with the MimetypesFileTypeMap in my code as well, which only considers the file extension. But that is something, that was previously insufficient in OpenMRS, simply because META-INF/mime.types, which is where the MimetypesFileTypeMap gets its mapping from, didn't contain many mimetypes. That is why I expanded META-INF/mime.types. You can take a look at the old mime.types and you will see that it's missing even some of the most basic mime types.

If you are against saving the mimetypes in filenames, that is also what I would suggest: set the mimetype via the MimetypesFileTypeMap when the obs is created, and set it the same way again when it is loaded from harddisk. This is still an improvement and more flexible than before. I think this is what you mean?

Reply:
Yes that is exactly what i mean! 👍

Reply:
I will go through the code and check what changes are necessary or have to be reverted.

Reply:
Will merge as soon as you are done with that. Thanks again for the great work and patience! 👍